components/ai_chat/core/browser/engine/oai_api_client.cc

```diff
@@ -0,0 +1,180 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/ai_chat/core/browser/engine/oai_api_client.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "base/functional/bind.h"
+#include "base/json/json_writer.h"
+#include "base/memory/weak_ptr.h"
+#include "base/strings/strcat.h"
+#include "base/strings/string_util.h"
+#include "base/types/expected.h"
+#include "brave/components/ai_chat/core/common/buildflags/buildflags.h"
+#include "brave/components/ai_chat/core/common/features.h"
+#include "brave/components/ai_chat/core/common/mojom/ai_chat.mojom.h"
+#include "brave/components/constants/brave_services_key.h"
+#include "net/http/http_request_headers.h"
+#include "net/http/http_status_code.h"
+#include "net/traffic_annotation/network_traffic_annotation.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+#include "url/gurl.h"
+
+namespace ai_chat {
+
+namespace {
+
+net::NetworkTrafficAnnotationTag GetNetworkTrafficAnnotationTag() {
+  return net::DefineNetworkTrafficAnnotation("ai_chat", R"(
+      semantics {
+        sender: "AI Chat"
+        description:
+          "This is used to communicate with Brave's AI Conversation API"
+          "on behalf of the user interacting with different browser AI"
+          "features."
+        trigger:
+          "Triggered by user interactions such as submitting an AI Chat"
+          "conversation message, or requesting a text rewrite."
+        data:
+          "Conversational messages input by the user as well as associated"
+          "content or user text to be rewritten. Can contain PII."
+        destination: WEBSITE
+      }
+      policy {
+        cookies_allowed: NO
+        policy_exception_justification:
+          "Not implemented."
+      }
+    )");
+}
+
+std::string CreateJSONRequestBody(base::Value::List messages,
+                                  const bool is_sse_enabled,
+                                  const mojom::CustomModel& model_config) {
+  base::Value::Dict dict;
+
+  dict.Set("messages", std::move(messages));
+  dict.Set("stream", is_sse_enabled);
+  dict.Set("temperature", 0.7);
+  dict.Set("model", model_config.model_request_name);
+
+  std::string json;
+  base::JSONWriter::Write(dict, &json);
+  return json;
+}
+
+}  // namespace
+
+OAIAPIClient::OAIAPIClient(
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory) {
+  api_request_helper_ = std::make_unique<api_request_helper::APIRequestHelper>(
+      GetNetworkTrafficAnnotationTag(), url_loader_factory);
+}
+
+OAIAPIClient::~OAIAPIClient() = default;
+
+void OAIAPIClient::ClearAllQueries() {
+  api_request_helper_->CancelAll();
+}
+
+void OAIAPIClient::PerformRequest(
+    const mojom::CustomModel& model_config,
+    base::Value::List messages,
```

Comment:
Consider accepting a custom type here (e.g. `struct OAIMessage {`}) - might make unit testing the Engine easier - because of the separation of the JSON parsing and reading, as well as being able to more easily construct and compare structs. At least, that's what I found with ConversationAPI.